docs/dependencies.md

```diff
@@ -0,0 +1,287 @@
+  of virtualenvs named by the hash of the relevant `requirements.txt`
+  file.  That way, when re-provisioning a development environment or
+  deploying a new production version with the same Python
+  dependencies, no downloading or installation is required: we just
+  use the same virtualenv.  When the only changes are upgraded
+  versions, we'll use [virtualenv-clone][] to clone the most similar
+  existing virtualenv and then just upgrade the packages needed,
+  making small version upgrades extremely efficient.  And finally, we
+  use `pip`'s built-in caching to ensure that a specific version of a
+  specific package is only downloaded once.
+* **Garbage-collecting caches**.  We have a tool,
+  `scripts/lib/clean-venv-cache`, which will clean old cached
+  virtualenvs that are no longer in use.  In production, the algorithm
+  preserves recent virtualenvs as well as those in use by any current
+  production deployment directory under `/home/zulip/deployments/`.
+  This helps ensure that a Zulip installation doesn't leak large
+  amounts of disk over time.
+* **Pinning versions of indirect depedendencies**.  We "pin" or "lock"
+  the versions of our indirect depedencies files with
+  `tools/update-locked-requirements` (powered by `pip-compile`).  What
+  this means is that we have some "source" requirements files, like
+  `requirements/common.txt`, that declare the packages that Zulip
+  depends on directly.  Those packages have their own recursive
+  dependencies.  When adding or removing a dependency from Zulip, one
+  simply edits the appropriate "source" requirements files, and then
+  runs `tools/update-locked-requirements`.  That tool will use `pip
+  compile` to generate the `prod_lock.txt` and `dev_lock.txt` files
+  that explicitly declare versions of all of Zulip's recursive
+  dependencies.  For indirect dependencies (i.e. dependencies not
+  explicitly declared in the source requirements files), it provides
+  helpful comments explaining which direct dependency (or
+  dependencies) needed that indirect dependency.  The process for
+  using this system is documented in more detail in
+  `requirements/README.md`.
+* **Scripts**.  Often, we want a script running in production to use
+  the Zulip virtualenv.  To make that work without a lot of duplicated
+  code, we have a helpful library,
+  `scripts/lib/setup_path_on_import.py`, which on import will put the
+  currently running Python script into the Zulip virtualenv.  This is
+  called by `./manage.py` to ensure that our Django code always uses
+  the correct virtualenv as well.
+
+## JavaScript and other frontend packages
+
+We use the same set of strategies described for Python dependencies
+for most of our JavaScript dependencies.  Consult the above section
+for details on the strategy; here we discuss how things differ.
+
+* **node_modules directory with pinned versions**.  We store
+  `node_modules` directories under `/srv/zulip-npm-cache` in a fashion
```

Comment:
Hmm, we don't mention `/srv/zulip-venv-cache`. I think mentioning the path is a good idea.